# Importing a library that is not in Colaboratory

To import a library that's not in Colaboratory by default, you can use `!pip install` or `!apt-get install`.

In [ ]:
!pip install matplotlib-venn

In [ ]:
!apt-get -qq install -y libfluidsynth1

E: Package 'libfluidsynth1' has no installation candidate


# Install 7zip reader [libarchive](https://pypi.python.org/pypi/libarchive)

In [ ]:
# https://pypi.python.org/pypi/libarchive
!apt-get -qq install -y libarchive-dev && pip install -U libarchive
import libarchive

Selecting previously unselected package libarchive-dev:amd64.
(Reading database ... 126284 files and directories currently installed.)
Preparing to unpack .../libarchive-dev_3.6.0-1ubuntu1.5_amd64.deb ...
Unpacking libarchive-dev:amd64 (3.6.0-1ubuntu1.5) ...
Setting up libarchive-dev:amd64 (3.6.0-1ubuntu1.5) ...
Processing triggers for man-db (2.10.2-1) ...
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 154.7/154.7 kB 5.4 MB/s eta 0:00:00
  Created wheel for libarchive: filename=libarchive-0.4.7-py3-none-any.whl size=31629 sha256=46c98cc7e1a5edc21ff00a2f808401826699ab4759f91330388efcb78fc7e953
  Stored in directory: /root/.cache/pip/wheels/32/98/bd/4893d6923dd027f455b250367d402bfd69a6f4416581df46db
Successfully built libarchive


# Install GraphViz & [PyDot](https://pypi.python.org/pypi/pydot)

In [ ]:
# https://pypi.python.org/pypi/pydot
!apt-get -qq install -y graphviz && pip install pydot
import pydot

# Install [cartopy](http://scitools.org.uk/cartopy/docs/latest/)

In [ ]:
!pip install cartopy
import cartopy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 27.3 MB/s eta 0:00:00


In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, pipeline
from datasets import load_dataset
import torch

# Hugging Face fitness veri setini yükle
veri_seti = load_dataset("chibbss/fitness-chat-prompt-completion-dataset", split="train")

# Kullanılacak model
model_adi = "teknium/OpenHermes-2.5-Mistral-7B"

# 4-bit quantization ayarları
quant_ayar = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4"
)

# Tokenizer ve model yüklemesi
tokenizer = AutoTokenizer.from_pretrained(model_adi, use_fast=True)
model = AutoModelForCausalLM.from_pretrained(
    model_adi,
    device_map="auto",
    quantization_config=quant_ayar
)

# Text-generation pipeline
sohbet = pipeline("text-generation", model=model, tokenizer=tokenizer)

# İlk 3 fitness örneğini test et
for i in range(3):
    soru = veri_seti[i]["prompt"]
    # OpenHermes formatına göre prompt
    giris = f"<|im_start|>user\n{soru}\n<|im_end|>\n<|im_start|>assistant\n"

    # Cevap üret
    sonuc = sohbet(giris, max_new_tokens=150, do_sample=True, temperature=0.7)[0]["generated_text"]
    cevap = sonuc.split("<|im_start|>assistant\n")[-1]

    print(f"🧍 Kullanıcı Sorusu: {soru}")
    print(f"🤖 OpenHermes Cevabı:\n{cevap.strip()}")
    print("=" * 60)


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


(…)ness-chat-prompt-completion-dataset.json: 0.00B [00:00, ?B/s]

Generating train split:   0%|          | 0/245 [00:00<?, ? examples/s]

PackageNotFoundError: No package metadata was found for bitsandbytes

In [ ]:
!pip install bitsandbytes

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 72.9/72.9 MB 11.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 48.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 42.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 20.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 6.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 12.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 8.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 6.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 57.9 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstallin

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, pipeline
from datasets import load_dataset
import torch

# Hugging Face fitness veri setini yükle
veri_seti = load_dataset("chibbss/fitness-chat-prompt-completion-dataset", split="train")

# Kullanılacak model
model_adi = "teknium/OpenHermes-2.5-Mistral-7B"

# 4-bit quantization ayarları
quant_ayar = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4"
)

# Tokenizer ve model yüklemesi
tokenizer = AutoTokenizer.from_pretrained(model_adi, use_fast=True)
model = AutoModelForCausalLM.from_pretrained(
    model_adi,
    device_map="auto",
    quantization_config=quant_ayar
)

# Text-generation pipeline
sohbet = pipeline("text-generation", model=model, tokenizer=tokenizer)

# İlk 3 fitness örneğini test et
for i in range(3):
    soru = veri_seti[i]["prompt"]
    # OpenHermes formatına göre prompt
    giris = f"<|im_start|>user\n{soru}\n<|im_end|>\n<|im_start|>assistant\n"

    # Cevap üret
    sonuc = sohbet(giris, max_new_tokens=150, do_sample=True, temperature=0.7)[0]["generated_text"]
    cevap = sonuc.split("<|im_start|>assistant\n")[-1]

    print(f"🧍 Kullanıcı Sorusu: {soru}")
    print(f"🤖 OpenHermes Cevabı:\n{cevap.strip()}")
    print("=" * 60)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


PackageNotFoundError: No package metadata was found for bitsandbytes

In [ ]:
!pip install bitsandbytes

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig, pipeline
from datasets import load_dataset
import torch

# Hugging Face fitness veri setini yükle
veri_seti = load_dataset("chibbss/fitness-chat-prompt-completion-dataset", split="train")

# Kullanılacak model
model_adi = "teknium/OpenHermes-2.5-Mistral-7B"

# 4-bit quantization ayarları
quant_ayar = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4"
)

# Tokenizer ve model yüklemesi
tokenizer = AutoTokenizer.from_pretrained(model_adi, use_fast=True)
model = AutoModelForCausalLM.from_pretrained(
    model_adi,
    device_map="auto",
    quantization_config=quant_ayar
)

# Text-generation pipeline
sohbet = pipeline("text-generation", model=model, tokenizer=tokenizer)

# İlk 3 fitness örneğini test et
for i in range(3):
    soru = veri_seti[i]["prompt"]
    # OpenHermes formatına göre prompt
    giris = f"<|im_start|>user\n{soru}\n<|im_end|>\n<|im_start|>assistant\n"

    # Cevap üret
    sonuc = sohbet(giris, max_new_tokens=150, do_sample=True, temperature=0.7)[0]["generated_text"]
    cevap = sonuc.split("<|im_start|>assistant\n")[-1]

    print(f"🧍 Kullanıcı Sorusu: {soru}")
    print(f"🤖 OpenHermes Cevabı:\n{cevap.strip()}")
    print("=" * 60)

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/493k [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/51.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/101 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/624 [00:00<?, ?B/s]

model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 2 files:   0%|          | 0/2 [00:00<?, ?it/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.94G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/4.54G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/120 [00:00<?, ?B/s]

Device set to use cuda:0


KeyError: 'prompt'

In [ ]:
# veri_seti'nin ilk öğesinin içeriğini görüntüle
print(veri_seti[0])

{'output': '1. Develop a consistent exercise routine – Exercise is essential for physical and mental health. Aim for at least 30 minutes of physical activity five days a week.\n\n2. Follow a healthy diet – Incorporate more fruits, vegetables, and whole grains into your diet while avoiding processed and fast foods.\n\n3. Get enough sleep – Give your body time to rest and repair by getting the recommended seven to nine hours of sleep every night.\n\n4. Practice relaxation techniques – Take a break to practice mindfulness, deep breathing, and other forms of relaxation to reduce stress and maintain emotional balance.\n\n5. Talk', 'instruction': 'What are some practical steps I can take to improve my overall health and well-being?'}


# Task
Calculate the BLEU and ROUGE scores for the model's output against the reference outputs in the dataset.

## Gerekli kütüphaneleri yükle

### Subtask:
BLEU ve ROUGE metriklerini hesaplamak için gerekli kütüphaneleri (NLTK ve rouge_score) yükleyin.


**Reasoning**:
The subtask requires installing NLTK and rouge_score libraries and downloading the 'punkt' resource from NLTK. This can be done in one code block.



In [ ]:
!pip install nltk rouge_score
import nltk
nltk.download('punkt')

  Preparing metadata (setup.py) ... done
  Created wheel for rouge_score: filename=rouge_score-0.1.2-py3-none-any.whl size=24934 sha256=ee1eeb9a55c26318c75125478c849245d794328c17d8993d6f1777f6ab6d5d61
  Stored in directory: /root/.cache/pip/wheels/1e/19/43/8a442dc83660ca25e163e1bd1f89919284ab0d0c1475475148
Successfully built rouge_score


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

## Modelden cevap üret

### Subtask:
Veri setindeki her "instruction" için modelden bir cevap üretin.


**Reasoning**:
Iterate through the dataset, generate responses for each instruction, and store the instruction and generated response.



In [ ]:
instructions_and_generated_answers = []

for example in veri_seti:
    instruction = example["instruction"]
    giris = f"<|im_start|>user\n{instruction}\n<|im_end|>\n<|im_start|>assistant\n"

    try:
        sonuc = sohbet(giris, max_new_tokens=150, do_sample=True, temperature=0.7)[0]["generated_text"]
        cevap = sonuc.split("<|im_start|>assistant\n")[-1]
        instructions_and_generated_answers.append((instruction, cevap.strip()))
    except Exception as e:
        print(f"Error processing instruction: {instruction}")
        print(f"Error: {e}")
        # Optionally append a placeholder or skip this example
        instructions_and_generated_answers.append((instruction, "Error generating response"))


print(f"Generated responses for {len(instructions_and_generated_answers)} instructions.")

Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
You seem to be using the pipelines sequentially on GPU. In order to maximize efficiency please use a dataset
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:32000 for open-end generation.
Setting `pad_token_id` to `

Generated responses for 245 instructions.


## Metrikleri hesapla

### Subtask:
Üretilen cevapları veri setindeki karşılık gelen "output" değerleriyle karşılaştırarak BLEU ve ROUGE skorlarını hesaplayın.


**Reasoning**:
Calculate the BLEU and ROUGE scores for each generated answer against its corresponding reference output and then compute the average scores.



In [ ]:
from nltk.translate.bleu_score import sentence_bleu
from rouge_score import rouge_scorer

reference_outputs = [example["output"] for example in veri_seti]
generated_answers = [answer for instruction, answer in instructions_and_generated_answers]

bleu_scores = []
rouge_scores = {
    "rouge-1": [],
    "rouge-2": [],
    "rouge-l": []
}

scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

for ref, gen in zip(reference_outputs, generated_answers):
    # Calculate BLEU score
    # BLEU expects a list of reference sentences, so we wrap the single reference in a list
    # It also expects tokenized sentences
    reference_tokens = [nltk.word_tokenize(ref)]
    generated_tokens = nltk.word_tokenize(gen)
    try:
        bleu = sentence_bleu(reference_tokens, generated_tokens)
        bleu_scores.append(bleu)
    except ZeroDivisionError:
        bleu_scores.append(0) # Append 0 if there's a division by zero (e.g., empty generated text)


    # Calculate ROUGE scores
    scores = scorer.score(ref, gen)
    rouge_scores["rouge-1"].append(scores["rouge1"].fmeasure)
    rouge_scores["rouge-2"].append(scores["rouge2"].fmeasure)
    rouge_scores["rouge-l"].append(scores["rougeL"].fmeasure)

# Calculate average scores
avg_bleu = sum(bleu_scores) / len(bleu_scores) if bleu_scores else 0
avg_rouge_1 = sum(rouge_scores["rouge-1"]) / len(rouge_scores["rouge-1"]) if rouge_scores["rouge-1"] else 0
avg_rouge_2 = sum(rouge_scores["rouge-2"]) / len(rouge_scores["rouge-2"]) if rouge_scores["rouge-2"] else 0
avg_rouge_l = sum(rouge_scores["rouge-l"]) / len(rouge_scores["rouge-l"]) if rouge_scores["rouge-l"] else 0

print(f"Average BLEU score: {avg_bleu}")
print(f"Average ROUGE-1 score: {avg_rouge_1}")
print(f"Average ROUGE-2 score: {avg_rouge_2}")
print(f"Average ROUGE-L score: {avg_rouge_l}")

LookupError: 
**********************************************************************
  Resource punkt_tab not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('punkt_tab')
  
  For more information see: https://www.nltk.org/data.html

  Attempted to load tokenizers/punkt_tab/english/

  Searched in:
    - '/root/nltk_data'
    - '/usr/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/share/nltk_data'
    - '/usr/local/share/nltk_data'
    - '/usr/lib/nltk_data'
    - '/usr/local/lib/nltk_data'
**********************************************************************


**Reasoning**:
The previous command failed because the 'punkt_tab' resource was not found. I need to download this resource using NLTK's downloader before proceeding with tokenization.



In [ ]:
import nltk
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

**Reasoning**:
Now that 'punkt_tab' is downloaded, I can re-run the code to calculate the BLEU and ROUGE scores.



In [ ]:
from nltk.translate.bleu_score import sentence_bleu
from rouge_score import rouge_scorer

reference_outputs = [example["output"] for example in veri_seti]
generated_answers = [answer for instruction, answer in instructions_and_generated_answers]

bleu_scores = []
rouge_scores = {
    "rouge-1": [],
    "rouge-2": [],
    "rouge-l": []
}

scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

for ref, gen in zip(reference_outputs, generated_answers):
    # Calculate BLEU score
    # BLEU expects a list of reference sentences, so we wrap the single reference in a list
    # It also expects tokenized sentences
    reference_tokens = [nltk.word_tokenize(ref)]
    generated_tokens = nltk.word_tokenize(gen)
    try:
        bleu = sentence_bleu(reference_tokens, generated_tokens)
        bleu_scores.append(bleu)
    except ZeroDivisionError:
        bleu_scores.append(0) # Append 0 if there's a division by zero (e.g., empty generated text)


    # Calculate ROUGE scores
    scores = scorer.score(ref, gen)
    rouge_scores["rouge-1"].append(scores["rouge1"].fmeasure)
    rouge_scores["rouge-2"].append(scores["rouge2"].fmeasure)
    rouge_scores["rouge-l"].append(scores["rougeL"].fmeasure)

# Calculate average scores
avg_bleu = sum(bleu_scores) / len(bleu_scores) if bleu_scores else 0
avg_rouge_1 = sum(rouge_scores["rouge-1"]) / len(rouge_scores["rouge-1"]) if rouge_scores["rouge-1"] else 0
avg_rouge_2 = sum(rouge_scores["rouge-2"]) / len(rouge_scores["rouge-2"]) if rouge_scores["rouge-2"] else 0
avg_rouge_l = sum(rouge_scores["rouge-l"]) / len(rouge_scores["rouge-l"]) if rouge_scores["rouge-l"] else 0

print(f"Average BLEU score: {avg_bleu}")
print(f"Average ROUGE-1 score: {avg_rouge_1}")
print(f"Average ROUGE-2 score: {avg_rouge_2}")
print(f"Average ROUGE-L score: {avg_rouge_l}")

/usr/local/lib/python3.11/dist-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/usr/local/lib/python3.11/dist-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/usr/local/lib/python3.11/dist-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 2-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_

Average BLEU score: 0.035339362807931915
Average ROUGE-1 score: 0.25161445379166997
Average ROUGE-2 score: 0.0642661495365208
Average ROUGE-L score: 0.13668142443401057


## Sonuçları görüntüle

### Subtask:
Hesaplanan BLEU ve ROUGE skorlarını görüntüleyin.


**Reasoning**:
Display the calculated average BLEU and ROUGE scores.



In [ ]:
print(f"Ortalama BLEU Skoru: {avg_bleu}")
print(f"Ortalama ROUGE-1 Skoru: {avg_rouge_1}")
print(f"Ortalama ROUGE-2 Skoru: {avg_rouge_2}")
print(f"Ortalama ROUGE-L Skoru: {avg_rouge_l}")

Ortalama BLEU Skoru: 0.035339362807931915
Ortalama ROUGE-1 Skoru: 0.25161445379166997
Ortalama ROUGE-2 Skoru: 0.0642661495365208
Ortalama ROUGE-L Skoru: 0.13668142443401057


## Summary:

### Data Analysis Key Findings

*   The average BLEU score for the model's generated text against the reference outputs is 0.0353.
*   The average ROUGE-1 score is 0.2516.
*   The average ROUGE-2 score is 0.0643.
*   The average ROUGE-L score is 0.1367.

### Insights or Next Steps

*   The low BLEU and ROUGE scores suggest that the model's generated text has limited n-gram overlap and lexical similarity with the reference outputs, indicating potential areas for improvement in text generation quality.
*   Further analysis could involve examining individual high and low-scoring examples to understand specific patterns of generation errors and inform model fine-tuning or prompt engineering strategies.
